class-plugin-autoupdate-filter.php

```diff
@@ -127,9 +130,49 @@ public function plugin_autoupdate_filter_custom_debug_email( $email, $failures,
 	 * @return string Customized HTML for automatic update settings.
 	 */
 	public function plugin_autoupdate_filter_custom_setting_html( $html, $plugin_file, $plugin_data ) {
+
+		// check if updates are explicitly blocked for this plugin
+		if ( function_exists( 'disable_autoupdate_specific_plugins' ) ) {
+
+			// create a fake object to feed to disable_autoupdate_specific_plugins
+			$plugin_obj                    = new stdClass();
+			$plugin_obj->slug              = dirname( $plugin_file );
+			$plugin_allowed_to_update_bool = disable_autoupdate_specific_plugins( true, $plugin_obj );
+
+			if ( false === $plugin_allowed_to_update_bool ) {
+				return 'Autoupdates have been explicitly deactivated for this plugin.';
+			}
+		}
+
 		return 'Automatic updates managed by <strong>Plugin Autoupdate Filter</strong>';
 	}
 
+	/**
+	 * Append text to upgrade text on plugins page for plugins explicitly set to not autoupdate
+	 *
+	 */
+	public function plugin_autoupdate_filter_change_upgrade_message_for_specific_plugins() {
+
+		$all_plugins = get_plugins();
+
+		foreach ( $all_plugins as $plugin_file => $plugin_data ) {
+			// create a fake object to feed to disable_autoupdate_specific_plugins
+			$plugin_obj                    = new stdClass();
+			$plugin_obj->slug              = dirname( $plugin_file );
+			$plugin_allowed_to_update_bool = disable_autoupdate_specific_plugins( true, $plugin_obj );
```

Comment:
This needs to be enclosed in a check to prevent a fatal error. I installed this on a test site and as soon as I activated it I got:

> **Fatal error**: Uncaught Error: Call to undefined function disable_autoupdate_specific_plugins() in /srv/users/user24d5787c/apps/user24d5787c/public/wp-content/plugins/plugin-autoupdate-filter/class-plugin-autoupdate-filter.php:162 Stack trace: #0 /srv/users/user24d5787c/apps/user24d5787c/public/wp-includes/class-wp-hook.php(310): Plugin_Autoupdate_Filter->plugin_autoupdate_filter_change_upgrade_message_for_specific_plugins(‘’) #1 /srv/users/user24d5787c/apps/user24d5787c/public/wp-includes/class-wp-hook.php(334): WP_Hook->apply_filters(NULL, Array) #2 /srv/users/user24d5787c/apps/user24d5787c/public/wp-includes/plugin.php(517): WP_Hook->do_action(Array) #3 /srv/users/user24d5787c/apps/user24d5787c/public/wp-admin/admin.php(175): do_action(‘admin_init’) #4 /srv/users/user24d5787c/apps/user24d5787c/public/wp-admin/plugins.php(10): require_once(‘/srv/users/user...’) #5 {main} thrown in **/srv/users/user24d5787c/apps/user24d5787c/public/wp-content/plugins/plugin-autoupdate-filter/class-plugin-autoupdate-filter.php** on line **162**

```suggestion
			if ( function_exists( 'disable_autoupdate_specific_plugins' ) ) {
				$plugin_allowed_to_update_bool = disable_autoupdate_specific_plugins( true, $plugin_obj );
			}
```

Even this creates a bunch of warnings:

```
Warning: Undefined variable $plugin_allowed_to_update_bool in /srv/users/user6bfc05ff/apps/user6bfc05ff/public/wp-content/plugins/plugin-autoupdate-filter/class-plugin-autoupdate-filter.php on line 165

Warning: Undefined variable $plugin_allowed_to_update_bool in /srv/users/user6bfc05ff/apps/user6bfc05ff/public/wp-content/plugins/plugin-autoupdate-filter/class-plugin-autoupdate-filter.php on line 165

Warning: Undefined variable $plugin_allowed_to_update_bool in /srv/users/user6bfc05ff/apps/user6bfc05ff/public/wp-content/plugins/plugin-autoupdate-filter/class-plugin-autoupdate-filter.php on line 165

Warning: Undefined variable $plugin_allowed_to_update_bool in /srv/users/user6bfc05ff/apps/user6bfc05ff/public/wp-content/plugins/plugin-autoupdate-filter/class-plugin-autoupdate-filter.php on line 165

Warning: Undefined variable $plugin_allowed_to_update_bool in /srv/users/user6bfc05ff/apps/user6bfc05ff/public/wp-content/plugins/plugin-autoupdate-filter/class-plugin-autoupdate-filter.php on line 165

Warning: Cannot modify header information - headers already sent by (output started at /srv/users/user6bfc05ff/apps/user6bfc05ff/public/wp-content/plugins/plugin-autoupdate-filter/class-plugin-autoupdate-filter.php:165) in /srv/users/user6bfc05ff/apps/user6bfc05ff/public/wp-admin/includes/misc.php on line 1431

Warning: Cannot modify header information - headers already sent by (output started at /srv/users/user6bfc05ff/apps/user6bfc05ff/public/wp-content/plugins/plugin-autoupdate-filter/class-plugin-autoupdate-filter.php:165) in /srv/users/user6bfc05ff/apps/user6bfc05ff/public/wp-includes/functions.php on line 6935

Warning: Cannot modify header information - headers already sent by (output started at /srv/users/user6bfc05ff/apps/user6bfc05ff/public/wp-content/plugins/plugin-autoupdate-filter/class-plugin-autoupdate-filter.php:165) in /srv/users/user6bfc05ff/apps/user6bfc05ff/public/wp-admin/admin-header.php on line 9

Warning: Cannot modify header information - headers already sent by (output started at /srv/users/user6bfc05ff/apps/user6bfc05ff/public/wp-content/plugins/plugin-autoupdate-filter/class-plugin-autoupdate-filter.php:165) in /srv/users/user6bfc05ff/apps/user6bfc05ff/public/wp-includes/option.php on line 1171

Warning: Cannot modify header information - headers already sent by (output started at /srv/users/user6bfc05ff/apps/user6bfc05ff/public/wp-content/plugins/plugin-autoupdate-filter/class-plugin-autoupdate-filter.php:165) in /srv/users/user6bfc05ff/apps/user6bfc05ff/public/wp-includes/option.php on line 1172
```

Reply:
Good catch! Fixed by using a return at the beginning of the function like this:

```php
if ( ! function_exists( 'disable_autoupdate_specific_plugins' ) ) {
return;
}
```

Also tested, and not getting warnings.